ConsoleApp for Newcommer

Patisson4 opened the pull request:
to review

Patisson4 committed:
ToString now uses GetType.Name

Patisson4 committed:
Merge branch 'master' into for_review

Patisson4 committed:
Update Program.cs

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+using System.Collections.ObjectModel;
+using System.IO;
 
+namespace ConsoleApp1
```
Даже тестовый проект нужно называть как-то более показательно)
Ещё очень важно: не вижу у тебя .csproj файла - файла с конфигурацией проекта. Когда мы пишем на C#, мы работаем с такими единицами, как класс, проект, решение. В решении может быть несколько проектов, в проекте несколько классов. По крайней мере без файла проекта твоё приложение не запустится 👀

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
 
+namespace ConsoleApp1
+{
+    internal class SubTask
```
Классы в C# по умолчанию и так `internal`, так что если ты решил сделать класс `internal` писать модификатор явно не обязательно. Ремарка на тему доступности класса: с какой целью ты хотел сделать класс `internal`? Так ли это важно? Если нет, то, может, стоит сделать класс `public`? Если ты создашь второй проект, который, условно, будет работать чисто как консольный принтер, то ты не сможешь получить доступ к `internal` классу оттуда, как-то так.

annchous replied:
Замечание к логике наименования класса: почему subtask? Это ведь **под**задача. Если бы я услышала такое название класса, то я могла бы предположить, что такое класс может быть наследником класса `Task`, однако в твоём случае это не так. Даже несмотря на иерархическую точку зрения класс подзадача без класса задача звучит не очень)

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+{
+    internal class SubTask
+    {
+        protected uint Id { get; }
```
Здорово, что пишешь `Id`, а не `ID`, как многие часто привыкли делать, но что не очень вяжется с неймингом в шарпах) Но тут у меня вопрос к `protected` модификатору: кто же будет тем счастливым наследником, которому будет предоставлен доступ к `Id`? Если ты рассчитывал сделать иерархию `Subtask : Task`, то тогда не забудь переименовать класс, если не планировал - то вполне можно сделать `Id` приватным (приватные поля именуются с нижнего подчёркивания, camel case'ом с маленькой буквы).

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+    internal class SubTask
+    {
+        protected uint Id { get; }
+        internal string TaskInfo { get; }
```
По поводу `internal` модификатора уже говорила)

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+        internal string TaskInfo { get; }
+        internal bool IsCompleted { get; private set; }
+
+        public SubTask(uint currentId, string info)
```
Вместо `currentId` можно просто `id`)

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+    {
+        public static void Main()
+        {
+            Manager l = new TaskManagerBase();
```
Можно упростить `Manager` до `var` и не нужно экономить на буквах (`l`) ))

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+
+    internal abstract class Manager
+    {
+        protected internal uint NextId { get; protected set; } = 0;
```
По умолчанию `uint` и так 0

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+            //add message needed? 
+            //Tasks.ContainsValue(value); ?
+            
+            var tsk = new SubTask(NextId, value);
```
Зачем сэкономил одну буковку и испортил слово =D Сделай `task`)

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+            
+            var tsk = new SubTask(NextId, value);
+            _tasks.TryAdd(NextId, tsk);  //what to do if task with next_id already exists?
+            NextId++;
```
Использовал `TryAdd`, но никак не воспользовался полезной информацией - возвращаемым значением этого метода. Что делать, если он вернёт `false` (и что это вообще означает)? Ты в любом случае увеличиваешь Id следующей задачи, но ведь если добавить задачу не удалось, то и `Id` не следует увеличивать.

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+            }
+        }
+
+        public override void Remove(uint id) //rename to Erase?
```
Не надо переименовывать, так ок)

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+            return buffer;
+        }
+
+        protected override void Parse(string rawData)
+        {
+            var data = rawData.Split('\n');
+            
+            //this = new TaskManagerBase();
+            _tasks = new SortedDictionary<uint, SubTask>();
+            
+            foreach (var line in data)
+            {
+                var statment = line.Split(' ');
+                if (string.Equals(statment[0], "<subtask"))
+                {
+                    Add(statment[5]);
+                    if (string.Equals(statment[3], "done"))
+                        Complete(NextId - 1);
+                }
+            }
+        }
```
Как-то неудобно так парсить, не так ли?) Советую прочитать про [json сериализацию](https://metanit.com/sharp/tutorial/6.5.php) и воспользоваться прочитанным)

annchous commented on ConsoleApp1/Program.cs:
```diff
@@ -1 +1,230 @@
+        
+        public override void ShowCompleted()
+        {
+            if (Convert.ToBoolean(CountCompleted))
```
В чём потребность сделать проверку на наличие выполненных задач именно методом конвертирования в `boolean` значение?) Почему не `CountCompleted > 0`?

annchous reviewed (changes requested):
Самое глобальное замечание: у нас (то есть у шарпистов) **один класс =** (практически всегда) **один файл**. Так что все свои классы распредели по отдельным `.cs` файлам с названием как у класса (`class Task` => `Task.cs`). Когда поправишь то, что я отметила, я ещё раз просмотрю твоё решение и, возможно, запрошу ещё некоторые изменения, удачи)

Patisson4 committed:
Refactor: Split all code into different files, minor changes

Patisson4 committed:
Add files via upload

Refactor: Split all code into different files, minor changes

Patisson4 committed:
Delete SuperTask.cs

Fix: wrong file deleted

Patisson4 committed:
Add files via upload

Patisson4 committed:
fix: wrong file deleted

Patisson4 committed:
fix: wrong file deleted

Patisson4 committed:
Delete ConsoleApp1 directory

directory name changed, content moved to another directory

Patisson4 committed:
Delete SuperTask.cs

commited earlier than it should, therefore file removed

Patisson4 committed:
Feat: Created ConsolePrinter
      Added new SuperTask class with deadline and subtask functionality
      Added indexing in Manager<T>
Refactor: Removed ToString() override in Task and SuperTask due to unnecessary
          Changes in command parsing system

Patisson4 committed:
feat: added .cproj file

Patisson4 committed:
fix: added Equals override for SuperTask

annchous commented on StartingTaskManagerProject/ConsolePrinter.cs:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+        
+        [Obsolete("Used for Newсomer")]
```
Довольно неожиданное применение этого атрибута. В принципе, ты понял его смысл, но, учитывая то, что в новой версии тестового задания отсутствует деление на Newcomer/Сектант, ненужный метод можно вообще убрать.

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+                    throw new IndexOutOfRangeException("Index of SortedDictionary is out of range");
+                return Tasks[i];
+            }
+        }
```
Вижу, что ты решил реализовать индексатор. На мой взгляд, здесь есть проблема, заключающаяся в том, что при использовании его в коде не очевидно, к чему ты обращаешься, пока не зайдёшь в код менеджера. Пример:
https://github.com/Patisson4/KysectStartupProject/blob/4308887dd6d226b950ad7cd26ddeb6723f3abf4d/StartingTaskManagerProject/Program.cs#L83
Могу предложить конкретизировать название менеджера, назвав его `TaskManager`

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+            }
+        }
+
+        public bool Add(T obj)
```
Обычно такие методы (которые возвращают булево значение при попытке совершения какого-то действия над коллекцией) называют с приставкой Try (например, `TryAdd`). Но это пока только к сведению. Раз у тебя только один метод по добавлению, можешь оставить так, чтобы не смущать пользователя.

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+
+        public bool Complete(uint id)
+        {
+            if (!Tasks.TryGetValue(id, out var buf))
```
Нейминг на `buf` плохой

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+            if (!Tasks.TryGetValue(id, out var buf))
+                return false;
+
+            if (buf.IsCompleted) //what to do if task already completed?
```
Ну комплитед и комплитед, чего бухтеть-то
Можешь не запариваться и оставить так

Patisson4 replied:
А стоит ли return убирать? Вроде бы и ускорение кода, а вроде лишний Complete() работает не настолько медленно, чтобы его ускорять

annchous replied:
но тогда он выполнит ещё и CountCompleted++

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+
+        public bool Remove(uint id)
+        {
+            if (!Tasks.Remove(id, out var buf))
```
Нейминг

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,158 @@
+            var manager = new Manager<SuperTask>();
+            string input;
+
+            //Ctrl+Z to successful stop
```
Кажется, этот коммент тут точно не к месту) Пользователь же не станет рыться в коде, чтобы понять, как остановить программу. Того, что у тебя есть команды через /, достаточно.

annchous commented on StartingTaskManagerProject/SuperTask.cs:
```diff
@@ -0,0 +1,24 @@
+        public DateTime? Deadline { get; } //should be made private?? i thought no cuz DataTime is immutable 
+        public Manager<Task> Subs { get; }
+        
+        public SuperTask(string taskInfo, DateTime? deadline = null, bool isCompleted = false, Manager<Task> subs = null) : base(taskInfo, isCompleted)
```
Ищи мой коммент про перегрузку для избавления от `= null` в аргументах

Patisson4 replied:
Для Manager есть проверка через ??
А Deadline проверяется на null при выводе чтобы не брать какую то конкретную дату за признак отсутствия дедлайна
Или SuperTask не может существовать без дедлайна?

annchous replied:
Ну, ты ведь можешь не делать эту проверку и дать пользователю явно понять, что этот параметр не нужен, если он не хочет его задавать сам)

annchous replied:
> Или SuperTask не может существовать без дедлайна?

Да по идее может

Patisson4 replied:
То же самое, конструктор по большей части для json сериализации

annchous commented on StartingTaskManagerProject/Task.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public class Task : BaseTask
+    {
+        public string TaskInfo { get; } //public for json
```
Коммент не нужен

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+{
+    public class Manager<T> where T : BaseTask
+    {
+        public uint NextId { get; private set; } //public for json
+        public uint CountCompleted { get; private set; } //public for json
+
+        public SortedDictionary<uint, T> Tasks { get; private set; } //public for json
+        [JsonIgnore] public ReadOnlyDictionary<uint, T> Data => new(Tasks); //for upper-level task usages
```
Комменты не нужны. Если они для тебя - это грустно, хотя бы ты должен понимать свой код без подсказок) Тем более он тут не громоздкий

Patisson4 replied:
Изначально все-таки большинство свойств хотелось сделать приватными во имя инкапсуляции (это же она? да? да?) Но поняв, что стандартному json десериализатору не справится без публичности этих свойств, с подачи Fredi поменял на public и оставил комментарии, описывающие причину

annchous replied:
> Изначально все-таки большинство свойств хотелось сделать приватными во имя инкапсуляции (это же она? да? да?)

Хех. Не сказала бы, что это исчерпывающее описание концепции инкапсуляции)

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+            var obj = JsonSerializer.Deserialize<Manager<T>>(file.ReadToEnd());
+
+            //this = obj
+            NextId = obj.NextId;
```
Случится катастрофа, если по какой-то причине `obj` останется null

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+        public void Load(string path)
+        {
+            var file = new StreamReader(path);
+            var obj = JsonSerializer.Deserialize<Manager<T>>(file.ReadToEnd());
```
нейминг

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,158 @@
+                                throw new ArgumentException(
+                                    "Method 'Save' must be invoked with exactly 1 argument");
+
+                            //check for valid path???
```
Можешь добавить в метод `Save` проверку. И убрать комментарий тут)

Patisson4 replied:
Тут вопрос вот в чем. Если у программы есть доступ, и название файла корректное, то файл создается. Я знаю как проверить доступность (или валидность, надо уточнить) пути к папке. Нужно ли проверять корректность названия файла и права доступа? (и есть ли вообще можно способы проверок этого?)

annchous replied:
А что ты подразумеваешь под проверкой корректности названия файла? Если имеется в виду расширение или ещё что-то, то пусть это останется на совести пользователя. Давай сойдёмся на том, что нас интересует только то, существует ли вообще путь к месту, которое указал пользователь для файла)

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,158 @@
+                        case "/add":
+                            switch (statment.Length)
+                            {
+                                case < 2:
```
Лучше это свалить в default, а конкретные случаи обработать нормально

Patisson4 commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,98 @@
+        public SortedDictionary<uint, T> Tasks { get; private set; } //public for json
+        [JsonIgnore] public ReadOnlyDictionary<uint, T> Data => new(Tasks); //for upper-level task usages
+
+        public Manager(uint nextId = 0, uint countCompleted = 0, SortedDictionary<uint, T> tasks = null)
```
Если я делаю там проверку на null с помощью ??, то можно не писать отдельный конструктор?
Или с ?? что то не так?
`Tasks = tasks ?? new SortedDictionary<uint, T>();`

annchous replied:
> Или с ?? что то не так?

Разве что то, что тебе приходится это делать, когда в общем-то и не обязательно) Где-то ниже (или выше) тоже затронула этот вопрос

Patisson4 replied:
Этот конструктор создан для работы json сериализации, пользователю он не должен быть интересен (в теории). Он ругается на отсутствие в конструкторе всех аргументов при вызове (скорее всего потому что сеттеры приватные, но их что-то не хочется делать публичными

annchous replied:
ладно

Patisson4 committed:
Refactoring and Minor fixes

Patisson4 committed:
Refactoring and Minor fixes

Patisson4 committed:
Feat: added GetHashCode override for both Task and SuperTask

Patisson4 committed:
Feat: added GetHashCode override for both Task and SuperTask

annchous commented on StartingTaskManagerProject/Manager.cs:
```diff
@@ -0,0 +1,97 @@
+            var file = new StreamReader(path);
+            var manager = JsonSerializer.Deserialize<Manager<T>>(file.ReadToEnd());
+
+            NextId = manager?.NextId ?? throw new Exception("Failed to deserialize in method Load");
```
Ну тут немного неочевидно, что ты проверяешь. К тому же это не единственное место использования десериализованного объекта. Я бы остановилась на проверке
```csharp
if (manager == null) throw...
```

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,183 @@
+                            if (!taskManager.Complete(completeId))
+                                throw new Exception("An error occured while marking object as complete");
+
+                            //if SuperTask is completed, then all its subtask must be also completed
```
А это, кстати, неправда (со слов автора тестового). Так что можно убрать

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,183 @@
+                            foreach (var (key, _) in taskManager[completeId].Subs.Data)
+                                taskManager[completeId].Subs.Complete(key);
+
+                            //check behaviour of previous for if Complete() will be changed
+
+                            Console.WriteLine("Object completed successfully");
```
Коммент убрать. Object обозвать чем-то более понятным (ну, Task, например?)

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,183 @@
+            var taskManager = new Manager<SuperTask>();
+            string input;
+
+            while ((input = Console.ReadLine()) != null && input != "") //input comparision may be redundant?
```
Возможно, чтобы не усложнять себе жизнь, можно сделать `while (true)` цикл и выходить из него только командой /exit и/или какой-то ещё. Ну и просто в начале каждой итерации считывать строку.

annchous commented on StartingTaskManagerProject/Program.cs:
```diff
@@ -0,0 +1,183 @@
+                        }
+
+                        case "/delete":
+                            //although method is called 'Remove', command is called 'Delete', that's why there is such a comment
```
Я бы догадалась и так, что к чему)

Patisson4 committed:
Fix: minor fixes
